Remap SRIOV interfaces

RamLavi opened the pull request:
**What this PR does / why we need it**:
This PR implements in steps the solution proposed on the SRIOV mapping [design](https://github.com/kubevirt/community/pull/185).

what is done so far (Commits):
- *sriov - Map sriov devices to pci addresses*
- *virt-controller, Patch network-pci map annotation*
- *virt-controller, Expose network-pci-map with DownwardAPI*
- *virt-launcher, sriov - Use new Network to PCI map with fallback*

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/kubevirt/kubevirt/issues/7444

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

rmohr commented:
@RamLavi could you describe the background of this change? Like why does the virt-launcher pod need this info and not for instance the VMI status?

ormergi commented on pkg/virt-controller/watch/virt-metadata/virt-metadata.go:
```diff
@@ -0,0 +1,82 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  * See the License for the specific language governing permissions and
```
nit: Move "* See the License for.." to a new line.

RamLavi replied:
Done

ormergi commented on pkg/virt-controller/watch/virt-metadata/virt-metadata.go:
```diff
@@ -0,0 +1,82 @@
+	}
+}
+
+func SetPodVirtMetadata(pod *v1.Pod, virtMetadata virtMeta) error {
+	jsonString, err := marshalToString(virtMetadata)
+	if err != nil {
+		return err
+	}
+
+	if len(pod.Annotations) == 0 {
+		pod.Annotations = make(map[string]string)
+	}
+
+	pod.Annotations[AnnotationKey] = jsonString
+	return nil
+}
+
+func GetPodVirtMetadata(pod *v1.Pod) (virtMeta, error) {
+	jsonString, ok := pod.Annotations[AnnotationKey]
+	if !ok {
+		return virtMeta{}, fmt.Errorf("%s key not found on pod annotations", AnnotationKey)
+	}
+
+	var virtMetadata virtMeta
+	err := json.Unmarshal([]byte(jsonString), &virtMetadata)
+	if err != nil {
+		return virtMeta{}, err
+	}
+
+	return virtMetadata, nil
+}
```
If I understand correctly this package is going to be used at virt-controller reconcile loop code,
I am not sure if those functions will be needed (maybe there are equivalents) or if this is the right place for them because it's not clear how it's going to be used.

Are you going to introduce how the virt-metadata package going to be used?

RamLavi replied:
well not in this PR but I was thinking something in the lines of 
```
func (c *VMIController) sync {
...
if !isTempPod(pod) && isPodReady(pod) {
...
virtMeta := virt_metadata.GetPodVirtMetadata(pod)
virtMeta.NetworkPciMap = DO_THE_MAP_THING()
virt_metadata.SetPodVirtMetadata(pod, virtMeta)
...
}
...
}
```

RamLavi replied:
@ormergi Can I resolve this?

ormergi commented on pkg/virt-controller/watch/virt-metadata/virt-metadata.go:
```diff
@@ -0,0 +1,82 @@
+	return virtMetadata, nil
+}
+
+func (m virtMeta) GetMarshaledString() string {
+	jsonString, _ := marshalToString(m)
+	return jsonString
+}
```
nit: I think it's easier to understand the code when an object's methods are closer to its definition/contractor.

RamLavi replied:
Done

ormergi commented on pkg/virt-controller/watch/virt-metadata/virt-metadata_test.go:
```diff
@@ -0,0 +1,114 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  * See the License for the specific language governing permissions and
```
nit: Move "* See the License for.." to a new line.

RamLavi replied:
Done

ormergi commented on pkg/virt-controller/watch/virt-metadata/virt-metadata.go:
```diff
@@ -0,0 +1,82 @@
+	return virtMetadata, nil
+}
+
+func (m virtMeta) GetMarshaledString() string {
```
nit: I would call in just String() or Json()/JsonString(), I think the "Get" prefix is redundant.

RamLavi replied:
totally agree. Done

ormergi reviewed (commented):
Hi Ram thanks for the PR, please see my inline comments.

ormergi commented:
> @RamLavi could you describe the background of this change? Like why does the virt-launcher pod need this info and not for instance the VMI status?

Currently, the mapping between VMI SR-IOV interfaces and the underlying VFs is inconsistent and the VMI may end up in a state each VMI interface is associated with the wrong VF.
The root cause for it is that there is not enough information at the time the mapping is performed.
The issue that tracking it is #7444.

ormergi commented:
@rmohr 
Currently, the mapping between VMI SR-IOV interfaces and the underlying VFs is not consistent where a VM might end up with SR-IOV interfaces that point to the wrong VFs with the wrong configurations (e.g: VLAN).
The root cause for it is the lack of information regarding the VFs PCI addresses when the mapping is performed.
The issue that tracks it is #7444.

rmohr commented:
> @rmohr
> Currently, the mapping between a VM SR-IOV interface and the underlying VFs is not consistent when more than one VF is from the same resource pool.
> The VM might end up with SR-IOV interfaces that point to the wrong VFs with the wrong configurations (e.g: VLAN).
> The root cause for it is the lack of information regarding the VFs PCI addresses when the mapping is performed.
> The issue that tracks it is #7444.

Does this info have to exist on the pod? Which entity needs it and when?

RamLavi commented:
> > @rmohr
> > Currently, the mapping between a VM SR-IOV interface and the underlying VFs is not consistent when more than one VF is from the same resource pool.
> > The VM might end up with SR-IOV interfaces that point to the wrong VFs with the wrong configurations (e.g: VLAN).
> > The root cause for it is the lack of information regarding the VFs PCI addresses when the mapping is performed.
> > The issue that tracks it is #7444.
> 
> Does this info have to exist on the pod? Which entity needs it and when?

Hey @rmohr, thanks to @ormergi explained the issue in short. Let me give you some more detail that may explain this change. If you have more questions, there is a more formal design on it that I'm sure @ormergi will be able to send.

- we want to set the mapping on virt-controller and use this data on virt-launcher [when](https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-launcher/virtwrap/device/hostdevice/sriov/hostdev.go#L36) creating the SRIOV devices. We will also support fall-back to the old mapping in case the new mapping fails to get enough info to map.
- How do we plan to map? the only way to correctly map the SRIOV networks to the PCI-address is to use the already existing network-status annotation. 
We however do not wish to expose all of it, as it may disclose too much of the cluster network topology to the guest user. Hence this PR was created, in order to expose only the mapping. (we plan to expose with DownwardAPI in a later PR)
Moreover, if other developers may find themselves in a situation where they need to expose other metadata, they will be able to, as this annotation is made so that it can accommodate that.

rmohr commented:
> Hey @rmohr, thanks to @ormergi explained the issue in short. Let me give you some more detail that may explain this change. If you have more questions, there is a more formal design on it that I'm sure @ormergi will be able to send.
> 
> * we want to set the mapping on virt-controller and use this data on virt-launcher [when](https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-launcher/virtwrap/device/hostdevice/sriov/hostdev.go#L36) creating the SRIOV devices. We will also support fall-back to the old mapping in case the new mapping fails to get enough info to map.
> * How do we plan to map? the only way to correctly map the SRIOV networks to the PCI-address is to use the already existing network-status annotation.
>   We however do not wish to expose all of it, as it may disclose too much of the cluster network topology to the guest user. Hence this PR was created, in order to expose only the mapping. (we plan to expose with DownwardAPI in a later PR)
>   Moreover, if other developers may find themselves in a situation where they need to expose other metadata, they will be able to, as this annotation is made so that it can accommodate that.

Thanks for the details. I am still not sure I fully understand the impact of all surrounding components. Do you intend to solve a problem just for KubeVirt or for a broader range? If so, the more natural place would probably be the VMI status, where you then automatically inherit the info in https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-launcher/virtwrap/device/hostdevice/sriov/hostdev.go#L36 on the VMI spec which we pass to virt-launcher.

rmohr commented:
> @ormergi correct me if I'm wrong, but I think the main issue with doing it on vmi-status is that we may get in trouble on migration. AFAIU, when migrating there is a period of time where both src and target pods run relating to the same vmi, meaning more than 1 entity (virt-controller for the src-pod and virt-handler for both pods) may have access to this information.
> Since the src-pod and target pod potentially hold different mappings, we decided it would be best to not use the vmi object as the place to put this info.
> Keeping things apart has a bonus benefit of keeping our future selves from using the wrong mapping in future changes.

While that is a good thought, I would still recommend exploring a more direct communication path. We have for multus exceptions where we work with annotations because that is just how the k8s integration works at this point (and clearly also a workaround since the community did not agree on how multi-network support should look like).

I guess you want  to implicitly say that these values differ on a target pod? One classical flow for that would be that both places can be represented on the VMI status, and that virt-handler sends only the info over which is interesting for the virt-launcher on the virt-handler node (look for VirtualMachineOptions which we send on virt-handler -> launcher sync along side the VMI). In addition, while I don't know if these values have anything to do with privileged operations at the moment, it avoids the pitfall that we have to look up downwards api values later from an untrusted virt-launcher to do privileged operations from virt-handler (virt-hanler can't see pods).

We sometimes have to take annotation shortcuts, but it always comes with architectural complications and tradeoffs on the virt-handler and virt-launcher communication side.

RamLavi commented:
> > @ormergi correct me if I'm wrong, but I think the main issue with doing it on vmi-status is that we may get in trouble on migration. AFAIU, when migrating there is a period of time where both src and target pods run relating to the same vmi, meaning more than 1 entity (virt-controller for the src-pod and virt-handler for both pods) may have access to this information.
> > Since the src-pod and target pod potentially hold different mappings, we decided it would be best to not use the vmi object as the place to put this info.
> > Keeping things apart has a bonus benefit of keeping our future selves from using the wrong mapping in future changes.
> 
> While that is a good thought, I would still recommend exploring a more direct communication path. We have for multus exceptions where we work with annotations because that is just how the k8s integration works at this point (and clearly also a workaround since the community did not agree on how multi-network support should look like).
> 
> I guess you want to implicitly say that these values differ on a target pod? One classical flow for that would be that both places can be represented on the VMI status, and that virt-handler sends only the info over which is interesting for the virt-launcher on the virt-handler node (look for VirtualMachineOptions which we send on virt-handler -> launcher sync along side the VMI). In addition, while I don't know if these values have anything to do with privileged operations at the moment, it avoids the pitfall that we have to look up downwards api values later from an untrusted virt-launcher to do privileged operations from virt-handler (virt-hanler can't see pods). We sometimes have to take annotation shortcuts, but it always comes with architectural complications and tradeoffs on the virt-handler and virt-launcher communication side.

I think we should properly discuss this on the design PR that @ormergi.
Until then, I'll move this PR to Draft.

EdDev commented:
> I think we should properly discuss this on the design PR that @ormergi will soon issue.
Until then, I'll move this PR to Draft.

Agreed.

Nevertheless, I still want to answer a few points below.
I hope to see the suggestions raised here put in the design as well, with the pros and cons to compare them between the various solutions.

> If KubeVirt, the more natural place would probably be the VMI status, where you then automatically inherit the info in https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-launcher/virtwrap/device/hostdevice/sriov/hostdev.go#L36 on the VMI spec which we pass to virt-launcher.

I would prefer not to see the information in the VMI status be used as a desired state input. Per my understanding of the pattern, spec is where the desired state is to be placed, status is where the current state should be exposed and objectmeta where extensions fit. Many times, data which has no good place is put in the objectmeta (e.g. data that no one really owns, but comes from a 3rd party source).

> While that is a good thought, I would still recommend exploring a more direct communication path.

We have explored multiple options. At this moment, the current proposal is the most direct one we found.
I guess it is our fault not to share the design proposal in parallel to this PR, it just changed a lot die to so many options on the table. We will post it very soon to expose the options that have been explored.

> I guess you want to implicitly say that these values differ on a target pod? 

Yes

> One classical flow for that would be that both places can be represented on the VMI status, and that virt-handler sends only the info over which is interesting for the virt-launcher on the virt-handler node (look for VirtualMachineOptions which we send on virt-handler -> launcher sync along side the VMI).

The option where the virt-handler is involved presented a complexity we identified as a significant cons.
Instead of two components being affected/aware of this information, 3 will need to be involved.

> In addition, while I don't know if these values have anything to do with privileged operations at the moment, it avoids the pitfall that we have to look up downwards api values later from an untrusted virt-launcher to do privileged operations from virt-handler (virt-hanler can't see pods).

The idea is to create an opertunity to pass information from the pod directly to the virt-launcher, not to let a priv entity to use it.
We can, and we should, make sure we warn from using this data by the virt-handler.
If the security is a concern, I am sure we can place enough warnings (and even linters) to protect against this.

> We sometimes have to take annotation shortcuts, but it always comes with architectural complications and tradeoffs on the virt-handler and virt-launcher communication side.

In this case, we think it it simplifies the architecture and avoid the need to couple irrelevant components from taking part in this interaction.

As a side note, the usage of this has been already used for the migration network.
So there is a precedence already, although a bit different.

I hope we can summarize all this useful information and options well in the design proposal in order to see the pros and cons of all the options.

rmohr commented:
> I would prefer not to see the information in the VMI status be used as a desired state input. Per my understanding of the pattern, spec is where the desired state is to be placed, status is where the current state should be exposed and objectmeta where extensions fit. 

This applies to specs of object creators. What we discuss here is just internal work. Status is a perfect place for that. Controllers just doing their work. Metadata is always the worst place for that. Without additional effort, it can not even be protected from user-modifications which requires required webhooks on pods (which we can't do, since then the whole cluster functionality would depend on our virt-api to be available) .

> Many times, data which has no good place is put in the objectmeta (e.g. data that no one really owns, but comes from a 3rd party source).

Yes, if you can't influence 3rd parties to a point you want, you sometimes have to place stuff in metadata. We don't have this situation here. We have established communication paths and the chance to use/extend them.

> The idea is to create an opertunity to pass information from the pod directly to the virt-launcher, not to let a priv entity to use it.
> We can, and we should, make sure we warn from using this data by the virt-handler.
> If the security is a concern, I am sure we can place enough warnings (and even linters) to protect against this.

Yes, that is not desirable for our architecture. It may even be secrity-wise ok for this specific set of data, but this direct communication flow is in general not wanted in our architecture. It is for instance worth having in virt-handler a switch which passes node-specific option to the current virt-launcher, in case that for some time for  multiple nodes exist. That is definitely not more complicated than establishing a new communication flow which involves downward api or env variables and reading files at additional arbitrary locations (we have enough of  that already), and then splitting the data up between two paths, one which goes through virt-handler to the converter and the other one going through the pod.

> In this case, we think it it simplifies the architecture and avoid the need to couple irrelevant components from taking part in this interaction.

The communication flows are already fully established, you don't have to add new components to the game due to the above reasons. We also already send all data through virt-handler and there is no need to protect virt-handler from that data be sent through.

RamLavi commented:
Example of virt-metadata on a pod with 3 SRIOV networks: {`sriovnet-vlan100-primary-mac`, `sriovnet-vlan100-secondary-mac`, `sriovnet-vlan100-third-mac`}:
```
virt-metadata: '{"network-pci-map":{"sriovnet-vlan100-primary-mac":"0000:65:01.2","sriovnet-vlan100-secondary-mac":"0000:65:01.3","sriovnet-vlan100-third-mac":"0000:65:01.7"}}'
```

RamLavi commented:
/test pull-kubevirt-e2e-kind-1.22-sriov

RamLavi commented:
/test pull-kubevirt-e2e-kind-1.22-sriov

moshe010 commented on pkg/network/multus/annotations.go:
```diff
@@ -0,0 +1,63 @@
+	v1 "kubevirt.io/api/core/v1"
+)
+
+func IterateMultusNonDefaultNetworks(vmi *v1.VirtualMachineInstance, fn func(network v1.Network, networkRequestInterfaceName string)) {
```
per of my comment on the spec this need to be change to support all networks. 
It better to look if network has k8s.v1.cni.cncf.io/resourceName in the annotation

RamLavi replied:
Hey @moshe010, do you mean the `k8s.v1.cni.cncf.io/resourceName` annotation on the NetworkAttachmentDefinition?
If so then I'm afraid it is not enough to directly map between the network and the pci-address.
Consider a situation where you have 2 networks (with different MACs) connected to the same NetworkAttachmentDefinition? in that case both will have the same k8s.v1.cni.cncf.io/resourceName (more info [here](https://github.com/kubevirt/community/pull/185/files#diff-aee927941b6b876c1ac6b0f0a74cac867f8112545776978799c0c065925f2c2eR285)).

We chose to rely on the multus pod annotation: `k8s.v1.cni.cncf.io/networks-status`. In it - you have a 1:1 mapping between the network name and the host-pci-address
To to the mapping we need to understand a few things:
1.  in the the `k8s.v1.cni.cncf.io/networks-status`, each network is distinguished by the`interface` field.
2.  The`interface` field is set from `k8s.v1.cni.cncf.io/networks` (which is created by kubevirt when creating the pod) 

`IterateMultusNonDefaultNetworks` is the function that makes sure that we give every network in vmi.spec.networks the appropriate and **consistent** name - so that we can rely on it when mapping the SRIOV networks via the `k8s.v1.cni.cncf.io/networks-status`

moshe010 replied:
when we use ovn-kubernetes with ovs hardware offload we use multus and NAD for primary network (even if there is not secanday network NAD)
 
if I understand you current logic,  you iterate on no default NAD (which mean secondary network). My comment was that you need to iterate also on the primary network which can be (NAD)  and handle it. 
update 

	for i, network := range multusNonDefaultNetworks {
                if network.default:
                   networkRequestInterfaceName := "eth0"
               else:
		   networkRequestInterfaceName := fmt.Sprintf("net%d", i+1)
		fn(network, networkRequestInterfaceName)
	}

and remove the filterMultusNonDefaultNetworks 

my comment regarding "k8s.v1.cni.cncf.io/resourceName" was incorrect because I don't understand what the code does.

RamLavi replied:
I see. can you paste a vmi and NAD manifests with the scenario you are referring to? it would make it easier for me to study and include this case.

moshe010 replied:
I am so sorry. I verified gain and we don't use primary network with kubevirt only with Pod. 
please discard all my comments on the support of primary network in kubevirt. your logic is good to assume secondary network only. Also please discard this from the spec proposal. 
Again sorry for the confusion

EdDev replied:
> when we use ovn-kubernetes with ovs hardware offload we use multus and NAD for primary network (even if there is not secanday network NAD)

@moshe010 , you are using the primary as a network device or as a vfio device?
At the moment we do not have this scenario covered in out e2e tests, and we do not map the primary network to a resource pool in the current implementation.
Said all that, I do not see the a reason not to add it to the mapping as well, in practice "fixing" an older limitation. But we better do it in a separate PR, one that we can reason for it with examples of usages.

RamLavi replied:
np @moshe010, thanks for letting me know. 
In any case, should the need for this arise in the future, it would be fairly easy to also map the primary network (but I agree that it should happen on a different PR).

moshe010 replied:
@EdDev @RamLavi 
So it seem there are 2 use case for VF on primary network:
1. VF as vfio device to pass though to the VM. currently it not support in kubevirt
2. VF as net-device which moved to Pod network namespace. The VF is connected to linux bridge and on the other end VM tap device. (this case is supported in kubevirt)
So for case 2 we still need to support primary network in the current code.

RamLavi commented:
/test pull-kubevirt-e2e-kind-1.22-sriov

RamLavi commented:
/test pull-kubevirt-e2e-kind-1.22-sriov

RamLavi commented:
/test pull-kubevirt-e2e-kind-1.22-sriov

RamLavi commented:
/test pull-kubevirt-e2e-kind-1.22-sriov

RamLavi commented:
/test pull-kubevirt-unit-test
/test pull-kubevirt-generate
/test pull-kubevirt-build
/test pull-kubevirt-e2e-k8s-1.24-sig-network

RamLavi commented:
/test pull-kubevirt-unit-test
/test pull-kubevirt-generate
/test pull-kubevirt-build
/test pull-kubevirt-e2e-k8s-1.24-sig-network

RamLavi commented:
/test pull-kubevirt-unit-test
/test pull-kubevirt-generate
/test pull-kubevirt-build
/test pull-kubevirt-e2e-k8s-1.24-sig-network

ormergi reviewed (commented):
Nice work!

I skipped the first 3 commits and reviewed the rest, I focused on the logic and a little bit of structure of the new code, see my in-line commits.

Other than that, there are commits that introduce code that is used on the following commit, please add a note to such commits that it's a preparation step/code that will be used later.

As a side note, I think there are functions that can be changed to not return an error and simplify the implementation a little bit (e.g: funcs that involve marshal/unmarshal operations).
We can discuss it offline and realize if it makes sense.

RamLavi commented on pkg/network/multus/network-pci-map.go:
```diff
@@ -0,0 +1,56 @@
+	k8sv1 "k8s.io/api/core/v1"
+)
+
+const NetworkPCIMapAnnot = "network-pci-map"
```
>How about network-to-pci-address-mapping or network-to-pci-address?

I prefer to keep it short and simple, "to" is redundant imo. 
I'm on the fence about "pci" vs "pci-address", I think it doesn't hurt the meaning (isn't it?). 

>Also, the change that adds kubevirt.io prefix should be moved to this commit.

totally!

RamLavi commented on pkg/network/multus/network-pci-map.go:
```diff
@@ -0,0 +1,56 @@
+	return networkPciMap
+}
+
+func (n NetworkPciMap) SetAnnotationInPod(pod *k8sv1.Pod) error {
+	jsonString, err := marshalToString(n)
+	if err != nil {
+		return err
+	}
+
+	if len(pod.Annotations) == 0 {
+		pod.Annotations = make(map[string]string)
+	}
+
+	pod.Annotations[NetworkPCIMapAnnot] = jsonString
+	return nil
+}
+
+func marshalToString(networkPciMap NetworkPciMap) (string, error) {
+	objBytes, err := json.Marshal(networkPciMap)
+	if err != nil {
+		return "", err
+	}
+	return string(objBytes), nil
+}
```
didn't catch your point - why is it only converting to string?

RamLavi commented on pkg/network/multus/network-pci-map.go:
```diff
@@ -0,0 +1,56 @@
+	return networkPciMap
+}
+
+func (n NetworkPciMap) SetAnnotationInPod(pod *k8sv1.Pod) error {
```
yep, Done

RamLavi commented on pkg/network/multus/network-pci-map_test.go:
```diff
@@ -0,0 +1,99 @@
+			networkPciMap:         nil,
+			expectedNetworkPciMap: nil,
+		}),
+		Entry("when networkPciMap is not nil, and annotation does not exist prior to set", addAnnotationToPodParams{
```
the annotation is not there when the pod is created, it is only added in virt-controller's Sync, and only if there are non-default SRIOV networks. So one scenario is that the annotation is not there prior to setting it.

RamLavi commented on pkg/network/sriov/sriov.go:
```diff
@@ -19,4 +19,84 @@
 const AliasPrefix = "sriov-"
+
+func PrepareSriovNetworkPciMap(vmi *virtv1.VirtualMachineInstance, pod *k8sv1.Pod) error {
+	networkPciMap, err := mapNetworkNameToPciAddress(vmi, pod)
+	if err != nil {
+		return err
+	}
+
+	if networkPciMap == nil {
+		return fmt.Errorf("network-pci map empty")
+	}
```
thinking aloud. 
if we don't get here if there are no SRIOV networks(using `SriovInterfaceExist()`), then I think you are right. Removing

RamLavi commented on pkg/network/sriov/sriov.go:
```diff
@@ -19,4 +19,84 @@
+		networkName := sriovIface.Name
+		multusInterfaceName := podInterfaceNamer[networkName]
+		networkStatusEntry := multusInterfaceNameToNetworkStatusMap[multusInterfaceName]
+		if networkStatusEntry.DeviceInfo == nil || networkStatusEntry.DeviceInfo.Pci == nil {
+			return nil, fmt.Errorf("SRIOV interface %s has no corresponding device-info/pci-address "+
+				"in the appropriate network status annotations", networkName)
+		}
+
+		pciAddress := networkStatusEntry.DeviceInfo.Pci.PciAddress
+		if pciAddress == "" {
+			return nil, fmt.Errorf("failed to associate pci-address to SRIOV interface %s", networkName)
+		}
+		networkPciMap[networkName] = pciAddress
+	}
+	return networkPciMap, nil
```
That's a good question.
Currently network-status annotation does not give partial data, but it may in the future (? maybe bug or some new feature).
But I think it may lead us to dark corners.. If there is no 1:1 mapping then what will the dom.xml look like? will it combust? We need to examine this and possible provide a hybrid solution, perhaps in a future PR..

ormergi replied:
>But I think it may lead us to dark corners.. If there is no 1:1 mapping then what will the dom.xml look like? will it combust?

The domain XML should be generated correctly, there will be 5 HostDevices objects, 3 of them with the correct PCI 
address, the other two might get the wrong PCI address (50% chance).
There should be no problem with sterilizing the HostDevices object to XML.

RamLavi commented on pkg/network/sriov/sriov_test.go:
```diff
@@ -0,0 +1,352 @@
+		Expect(err).To(HaveOccurred(), "should fail getting network-pci-map from pod")
+		Expect(networkPCIMap).To(BeNil())
+	},
+		Entry("when pod's networkStatus Annotation is invalid",
```
good catch! totally

RamLavi commented:
/test pull-kubevirt-unit-test
/test pull-kubevirt-generate
/test pull-kubevirt-build
/test pull-kubevirt-e2e-k8s-1.24-sig-network
/test pull-kubevirt-e2e-kind-1.22-sriov

RamLavi commented:
/test pull-kubevirt-unit-test
/test pull-kubevirt-generate
/test pull-kubevirt-build
/test pull-kubevirt-e2e-k8s-1.24-sig-network
/test pull-kubevirt-e2e-kind-1.22-sriov

RamLavi commented:
/test pull-kubevirt-unit-test
/test pull-kubevirt-generate
/test pull-kubevirt-build
/test pull-kubevirt-e2e-k8s-1.24-sig-network
/test pull-kubevirt-e2e-kind-1.22-sriov

RamLavi commented:
/test pull-kubevirt-unit-test
/test pull-kubevirt-generate
/test pull-kubevirt-build
/test pull-kubevirt-e2e-k8s-1.24-sig-network
/test pull-kubevirt-e2e-kind-1.22-sriov

fails seem unrelated

RamLavi commented on pkg/virt-controller/watch/vmi.go:
```diff
@@ -1059,6 +1088,18 @@ func (c *VMIController) sync(vmi *virtv1.VirtualMachineInstance, pod *k8sv1.Pod,
 
 	if !isTempPod(pod) && isPodReady(pod) {
+		if sriov.SriovInterfaceExist(vmi) {
+			err := sriov.PrepareSriovNetworkPciMap(vmi, pod)
```
I thought I addressed this, sorry

I agree it's ugly. I think we should patch the pod (and also the vmi) after we do the `Sync()` + `updateStatus()` code.
I'll try to do it in a separate PR, as it is irrelevant to this one. 
But I don't think `PrepareSriovNetworkPciMap` should hide behind a general `syncPodAnnotations` function.

RamLavi commented:
/test pull-kubevirt-unit-test
/test pull-kubevirt-generate
/test pull-kubevirt-build
/test pull-kubevirt-e2e-k8s-1.24-sig-network
/test pull-kubevirt-e2e-kind-1.22-sriov

RamLavi commented:
/test pull-kubevirt-unit-test
/test pull-kubevirt-generate
/test pull-kubevirt-build
/test pull-kubevirt-e2e-k8s-1.24-sig-network
/test pull-kubevirt-e2e-kind-1.22-sriov

RamLavi commented:
/test pull-kubevirt-unit-test
/test pull-kubevirt-generate
/test pull-kubevirt-build
/test pull-kubevirt-e2e-k8s-1.24-sig-network
/test pull-kubevirt-e2e-kind-1.22-sriov

ormergi commented on pkg/virt-launcher/virtwrap/device/hostdevice/sriov/pcipool-network-status.go:
```diff
@@ -0,0 +1,77 @@
+	return pool, nil
+}
+
+func (p *PCIAddressWithNetworkStatusPool) loadVirtMetadataVolume(networkPciMapPath string) (map[string]string, error) {
```
This should be renamed to reflect the new annotation name.

RamLavi replied:
Done

ormergi commented on pkg/virt-controller/watch/vmi.go:
```diff
@@ -1056,6 +1086,21 @@ func (c *VMIController) sync(vmi *virtv1.VirtualMachineInstance, pod *k8sv1.Pod,
+				pod.Annotations[sriov.NetworkPCIMapAnnot] = networkPciMapAnnotationValue
+				err = c.syncPodAnnotations(pod)
+				if err != nil {
+					return &syncErrorImpl{fmt.Errorf("failed to sync pod annotations: %v", err), FailedPodPatchReason}
```
I think this error should be close to the Patch operation that failed instead.
Or move the patch operation to be inlined "here" (at sync()).

RamLavi replied:
changed to only return the err message

ormergi commented on pkg/virt-controller/watch/vmi.go:
```diff
@@ -1056,6 +1086,21 @@ func (c *VMIController) sync(vmi *virtv1.VirtualMachineInstance, pod *k8sv1.Pod,
+				vmi.Spec.Networks, vmi.Spec.Domain.Devices.Interfaces, pod.Annotations[networkv1.NetworkStatusAnnot],
+			)
+			if err != nil {
+				log.Log.V(3).Object(vmi).Warningf("failed to prepare SR-IOV network-pci map %q", err)
```
Does it make sense to make the user aware of it (SR-IOV mapping might be wrong)?
If so, it should be raised as a warning event instead of a log message.

RamLavi replied:
removed the V(3)

ormergi commented on pkg/virt-controller/services/rendervolumes.go:
```diff
@@ -54,6 +55,7 @@ func (vr *VolumeRenderer) Mounts() []k8sv1.VolumeMount {
 		mountPathWithPropagation(containerDisks, vr.containerDiskDir, k8sv1.MountPropagationHostToContainer),
 		mountPath("libvirt-runtime", "/var/run/libvirt"),
 		mountPath("sockets", filepath.Join(vr.virtShareDir, "sockets")),
+		mountPath(sriov.VolumeName, sriov.MountPath),
```
Note that this mount will be added to every pod.

RamLavi replied:
Fixed

EdDev commented on pkg/virt-controller/watch/vmi.go:
```diff
@@ -345,7 +349,22 @@ func (c *VMIController) execute(key string) error {
+func (c *VMIController) syncPodAnnotations(updatedPod *k8sv1.Pod) error {
+	patchBytes, err := prepareAnnotationsPatch(updatedPod.Annotations)
+	if err != nil {
+		return fmt.Errorf("failed to prepare patch pod: %v", err)
```
I think this info needs to be added in `prepareAnnotationsPatch`.

RamLavi replied:
moved

RamLavi commented:
/test pull-kubevirt-unit-test
/test pull-kubevirt-generate
/test pull-kubevirt-build
/test pull-kubevirt-e2e-k8s-1.24-sig-network
/test pull-kubevirt-e2e-kind-1.22-sriov

RamLavi commented on pkg/network/vmispec/interface_test.go:
```diff
@@ -65,6 +65,7 @@ var _ = Describe("VMI network spec", func() {
 			}
 
 			Expect(netvmispec.FilterSRIOVInterfaces(ifaces)).To(BeEmpty())
+			Expect(netvmispec.SriovInterfaceExist(ifaces)).To(BeFalse())
```
I admit it's a classic piggy back, bu why not? are we missing a scenario we want to test?

RamLavi commented on pkg/network/sriov/sriov.go:
```diff
@@ -19,4 +19,74 @@
+	networkPciMap := map[string]string{}
+	for _, sriovIface := range vmispec.FilterSRIOVInterfaces(interfaces) {
+		multusInterfaceName := networkNameScheme[sriovIface.Name]
+		networkStatusEntry := multusInterfaceNameToNetworkStatusMap[multusInterfaceName]
```
IIUC `k8s.v1.cni.cncf.io/network-status` creates an entry for each entry in `k8s.v1.cni.cncf.io/networks`
Since we explicitly create the entries of  `k8s.v1.cni.cncf.io/networks` [here](https://github.com/kubevirt/kubevirt/pull/8226/files#diff-d317cef6f369fd934b3c1086561948ea5a267dc3e5beaf38898be013d088a01aL59) using the same nameScheme function - then we are guaranteed that `multusInterfaceNameToNetworkStatusMap[multusInterfaceName]` exists, since 
multusInterfaceName is a sriovIface.Name which is a subset of the SRIOV networks...

RamLavi replied:
but we can always have bugs on Multus so I'll address this issue

RamLavi replied:
Done. Will now fail if not any SRIOV networks is not showing on the network-status list

RamLavi commented on pkg/network/vmispec/interface_test.go:
```diff
@@ -87,6 +88,7 @@ var _ = Describe("VMI network spec", func() {
 			}
 
 			Expect(netvmispec.FilterSRIOVInterfaces(ifaces)).To(Equal([]v1.Interface{sriov_net1, sriov_net2}))
+			Expect(netvmispec.SriovInterfaceExist(ifaces)).To(BeTrue())
```
ditto

RamLavi commented on pkg/virt-controller/watch/vmi.go:
```diff
@@ -131,6 +133,8 @@ const (
 	// is suitable for migration (since CPU model / required features are not supported)
 	NoSuitableNodesForHostModelMigration = "NoSuitableNodesForHostModelMigration"
+	// FailedPodPatchReason is set when a pod patch error occurs during sync
+	FailedPodPatchReason = "FailedPodPatchSync"
```
Done

RamLavi commented on pkg/virt-controller/services/rendervolumes.go:
```diff
@@ -67,6 +69,8 @@ func (vr *VolumeRenderer) Volumes() []k8sv1.Volume {
 		emptyDirVolume("ephemeral-disks"),
 		emptyDirVolume(containerDisks),
+		//Fixme data processed on virt-controller should ideally pass through virt-hanlder gRPC API.
+		downwardAPIDirVolume(sriov.VolumeName, sriov.VolumePath, fmt.Sprintf("metadata.annotations['%s']", sriov.NetworkPCIMapAnnot)),
```
I remembered that it didn't but I double checked and it does.. We may want to incorporate `isSRIOVExist()` ... WDYT?

RamLavi replied:
> Thank you for the PR @RamLavi. I've reviewed most of the first commit.

@orelmisan I'm afraid this commit is related to the consolidate PR 

> Thank you for the PR @RamLavi. I've reviewed most of the first commit.

@orelmisan I'm afraid this commit is part of the former PR https://github.com/kubevirt/kubevirt/pull/8263, as it is mentioned on this PR desc.

RamLavi replied:
Fixed

RamLavi commented on pkg/network/sriov/sriov.go:
```diff
@@ -19,4 +19,74 @@
 const AliasPrefix = "sriov-"
+
+func CreateNetworkPCIAnnotationValue(networks []virtv1.Network, interfaces []virtv1.Interface, networkStatusAnnotationValue string) (string, error) {
+	networkPciMap, err := mapNetworkNameToPciAddress(networks, interfaces, networkStatusAnnotationValue)
```
Done

RamLavi commented on pkg/network/sriov/sriov.go:
```diff
@@ -19,4 +19,74 @@
+	return string(networkPciMapBytes), nil
+}
+
+func mapNetworkNameToPciAddress(networks []virtv1.Network, interfaces []virtv1.Interface, networkStatusAnnotationValue string) (map[string]string, error) {
```
DONE

RamLavi commented on pkg/network/sriov/sriov_test.go:
```diff
@@ -0,0 +1,302 @@
+			},
+			networkStatusWithTwoSRIOVNetworksButOneWithNoDeviceInfoData,
+		),
+	)
```
Added such a scenario. Done

RamLavi commented on pkg/virt-controller/watch/vmi.go:
```diff
@@ -345,7 +349,22 @@ func (c *VMIController) execute(key string) error {
 	return nil
+}
+
+func (c *VMIController) syncPodAnnotations(updatedPod *k8sv1.Pod) error {
```
changed this function to accept only new annotations, and set them appropriately.

RamLavi commented on manifests/generated/operator-csv.yaml.in:
```diff
@@ -503,6 +503,7 @@ spec:
           - delete
           - update
           - create
+          - patch
```
strange.. but without it my patch did not work..

RamLavi commented on pkg/network/vmispec/interface.go:
```diff
@@ -39,6 +39,15 @@ func FilterSRIOVInterfaces(ifaces []v1.Interface) []v1.Interface {
 	return sriovIfaces
 }
 
+func SriovInterfaceExist(ifaces []v1.Interface) bool {
```
Done

RamLavi commented on pkg/virt-controller/watch/vmi.go:
```diff
@@ -345,7 +349,22 @@ func (c *VMIController) execute(key string) error {
+	}
 
+	if len(patchBytes) > 0 {
+		_, err = c.clientset.CoreV1().Pods(updatedPod.Namespace).Patch(context.TODO(), updatedPod.Name, types.JSONPatchType, patchBytes, v1.PatchOptions{})
```
Done

RamLavi commented on pkg/virt-controller/watch/vmi.go:
```diff
@@ -345,7 +349,22 @@ func (c *VMIController) execute(key string) error {
+		_, err = c.clientset.CoreV1().Pods(updatedPod.Namespace).Patch(context.TODO(), updatedPod.Name, types.JSONPatchType, patchBytes, v1.PatchOptions{})
+		if err != nil {
+			log.Log.Object(updatedPod).Errorf("Patching of pod failed during sync: %v", err)
+			return fmt.Errorf("failed to patch pod: %v", err)
```
changed to return only err

RamLavi commented:
@ormergi @EdDev addressed all comments [here](https://github.com/kubevirt/kubevirt/compare/604cacfba3c92cbff03720448fd71c6fd704d804..434a58890b92373e34471baa75e7a94b8f7e2388) (except the dynamic downwardAPI behavior which I still need to do)

RamLavi commented:
/test pull-kubevirt-unit-test
/test pull-kubevirt-generate
/test pull-kubevirt-build
/test pull-kubevirt-e2e-k8s-1.24-sig-network
/test pull-kubevirt-e2e-kind-1.22-sriov

RamLavi commented:
/test pull-kubevirt-e2e-kind-1.22-sriov

RamLavi commented:
/test pull-kubevirt-e2e-kind-1.22-sriov

orelmisan commented:
There is a small typo in the PR description:
> What this PR does / why we need it:
This PR implements in steps the solution proposed on the SRIOV mapping **desgin**.

Should be "design".

orelmisan commented on pkg/network/sriov/sriov.go:
```diff
@@ -19,4 +19,77 @@
+
+	networkPciMap := map[string]string{}
+	for _, sriovIface := range vmispec.FilterSRIOVInterfaces(interfaces) {
+		multusInterfaceName := networkNameScheme[sriovIface.Name]
```
```suggestion
		podIfaceName := networkNameScheme[sriovIface.Name]
```

RamLavi replied:
I think we should keep this. although it's the pod interface name, in our context it is the multus network-status annotation "interface" field...

orelmisan commented on pkg/network/sriov/sriov.go:
```diff
@@ -19,4 +19,77 @@
+	}
+	networkNameScheme := namescheme.CreateNetworkNameScheme(networks)
+
+	networkPciMap := map[string]string{}
```
This map is called `networkPciMap` but it actually maps VMI interface names to PCI addresses.
```suggestion
	ifaceNameToPciAddrMap := map[string]string{}
```

RamLavi replied:
it's not. specifically it maps names from vmi.spec.networks[i].Name to PCI-Addresses

orelmisan replied:
The loop bellow is iterating over `vmispec.FilterSRIOVInterfaces(interfaces)` and takes their name `sriovIface.Name` which is linking to the network name.
I find this a bit confusing.

RamLavi replied:
yes, the interface.name is linking to the network.name.
It's confusing because vmi spec is confusing by design.. 

The reason I am iterating the interfaces is because only interfaces have the "sriov" field.
However the "interface" filed in the multus netowrk status annotation (i.e. net1", "net2"..) is indexed from the network list, and is connected directly to it's order. The link between interfaces and networks is the name (interface.Name points to the network name)..
I'm afraid the confusion is by design..

orelmisan commented on pkg/network/sriov/sriov.go:
```diff
@@ -19,4 +19,77 @@
+}
+
+func mapNetworkNameToPCIAddress(networks []virtv1.Network, interfaces []virtv1.Interface, networkStatusAnnotationValue string) (map[string]string, error) {
+	multusInterfaceNameToNetworkStatusMap, err := mapMultusInterfaceNameToNetworkStatus(networkStatusAnnotationValue)
```
```suggestion
	podIfaceNameToNetworkStatusMap, err := mapMultusInterfaceNameToNetworkStatus(networkStatusAnnotationValue)
```

orelmisan commented on pkg/network/sriov/sriov.go:
```diff
@@ -19,4 +19,77 @@
+		multusInterfaceName := networkNameScheme[sriovIface.Name]
+		networkStatusEntry, exist := multusInterfaceNameToNetworkStatusMap[multusInterfaceName]
+		if !exist {
+			return nil, fmt.Errorf("failed to find network-status entry with interface %s", multusInterfaceName)
```
```suggestion
			return nil, fmt.Errorf("failed to find network-status entry with Pod interface %q", multusInterfaceName)
```

RamLavi replied:
again - in this context the pod interface name is also used as the network-status "interface" field..

RamLavi replied:
why do we need to use %q? we are not attaching the error

orelmisan replied:
`%q` in this context makes the `multusInterfaceName` appear between quotes.
https://pkg.go.dev/fmt
So the error message will be for example:
> failed to find network-status entry with Pod interface "net1"

https://pkg.go.dev/fmt

RamLavi replied:
ah, didn't know that, cool! :)

RamLavi replied:
fixed throughout

orelmisan commented on pkg/network/sriov/sriov.go:
```diff
@@ -19,4 +19,77 @@
+			return nil, fmt.Errorf("failed to find network-status entry with interface %s", multusInterfaceName)
+		}
+		if networkStatusEntry.DeviceInfo == nil || networkStatusEntry.DeviceInfo.Pci == nil {
+			return nil, fmt.Errorf("failed to find device-info/pci-address in network-status annotation for SR-IOV interface %s", sriovIface.Name)
```
```suggestion
			return nil, fmt.Errorf("failed to find device-info/ PCI address in network-status annotation for SR-IOV interface %q", sriovIface.Name)
```

RamLavi replied:
why do we need to use %q? 
fixed the "PCI"

RamLavi replied:
fixed the %q

orelmisan commented on pkg/network/sriov/sriov.go:
```diff
@@ -19,4 +19,77 @@
+
+		pciAddress := networkStatusEntry.DeviceInfo.Pci.PciAddress
+		if pciAddress == "" {
+			return nil, fmt.Errorf("failed to associate pci-address to SRIOV interface %s", sriovIface.Name)
```
```suggestion
			return nil, fmt.Errorf("failed to associate PCI address to SR-IOV interface %q", sriovIface.Name)
```

RamLavi replied:
ditto

RamLavi replied:
fixed the %q

orelmisan commented on pkg/network/sriov/sriov.go:
```diff
@@ -19,4 +19,77 @@
+	return networkPciMap, nil
+}
+
+func mapMultusInterfaceNameToNetworkStatus(networkStatusAnnotationValue string) (map[string]networkv1.NetworkStatus, error) {
```
```suggestion
func mapPodIfaceNameToNetworkStatus(networkStatusAnnotationValue string) (map[string]networkv1.NetworkStatus, error) {
````

RamLavi replied:
ditto

orelmisan commented on pkg/network/sriov/sriov.go:
```diff
@@ -19,4 +19,77 @@
+		return nil, fmt.Errorf("network-status annotation is not present")
+	}
+	var networkStatusList []networkv1.NetworkStatus
+	err := json.Unmarshal([]byte(networkStatusAnnotationValue), &networkStatusList)
+	if err != nil {
```
```suggestion
	if err := json.Unmarshal([]byte(networkStatusAnnotationValue), &networkStatusList); err != nil {
```

RamLavi replied:
I find this format less appealing. It shortens the lines but also lengthens the it, do you care to explain why this is preferred?

orelmisan replied:
It limits the `err` variable's scope to the `if` statement.

RamLavi replied:
fixed.

orelmisan commented on pkg/network/sriov/sriov.go:
```diff
@@ -19,4 +19,77 @@
+		return nil, fmt.Errorf("failed to unmarshal network-status annotation: %q", err)
+	}
+
+	multusInterfaceNameToNetworkStatusMap := map[string]networkv1.NetworkStatus{}
```
```suggestion
	podIfaceNameToNetworkStatusMap := map[string]networkv1.NetworkStatus{}
```

RamLavi replied:
ditto

orelmisan reviewed (commented):
Thank you for the PR @RamLavi.
I've reviewed most of the first commit.

RamLavi commented:
@EdDev @ormergi fix the patching and added a condintioned DownwarAPI approach [here](https://github.com/kubevirt/kubevirt/compare/0c2f3efb8a3ce907804251802ed471b226d841dd..9aae7dc0f00f9eff6e1c833cb90cc4141a69032c)

RamLavi commented:
/test pull-kubevirt-unit-test
/test pull-kubevirt-generate
/test pull-kubevirt-build
/test pull-kubevirt-e2e-k8s-1.24-sig-network
/test pull-kubevirt-e2e-kind-1.22-sriov

RamLavi commented on pkg/network/sriov/sriov.go:
```diff
@@ -19,4 +19,77 @@
+
 const AliasPrefix = "sriov-"
+
+func CreateNetworkPCIAnnotationValue(networks []virtv1.Network, interfaces []virtv1.Interface, networkStatusAnnotationValue string) (string, error) {
```
I prefer the one I used. `Create` is the correct in this context and `networkStatusAnnotation` is a used term (i.e. the network-Status annotation that multus used and we digest to create our map)

orelmisan replied:
My suggestion was to drop the `Value` suffix.
I don't see a problem with the verb `Create` or the `networkStatusAnnotation` name.

RamLavi commented on pkg/network/sriov/sriov.go:
```diff
@@ -19,4 +19,77 @@
+		return "", err
+	}
+
+	if len(networkPciMap) == 0 {
+		return "", fmt.Errorf("expecting non-emtpy network-pci-map")
+	}
```
now I'm struggling with this one.. If we "wrap" this function with the "sriovInterfaceExists()" function then no, but leaving it exposed feels wrong. WDYT?

RamLavi replied:
changed, tell me what do you think :)

orelmisan replied:
I don't find this code anymore (in the first commit), was it removed?

RamLavi commented:
[Rebased](https://github.com/kubevirt/kubevirt/compare/9aae7dc0f00f9eff6e1c833cb90cc4141a69032c..156eff732dfccf85eaf73b4ce96d9196e797ba4e)

RamLavi commented:
/test pull-kubevirt-unit-test
/test pull-kubevirt-generate
/test pull-kubevirt-build
/test pull-kubevirt-e2e-k8s-1.24-sig-network
/test pull-kubevirt-e2e-kind-1.22-sriov

RamLavi commented:
/test pull-kubevirt-unit-test
/test pull-kubevirt-generate
/test pull-kubevirt-build
/test pull-kubevirt-e2e-k8s-1.24-sig-network
/test pull-kubevirt-e2e-kind-1.22-sriov

RamLavi commented on pkg/network/sriov/sriov.go:
```diff
@@ -19,4 +19,77 @@
+	if err != nil {
+		return nil, err
+	}
+	networkNameScheme := namescheme.CreateNetworkNameScheme(networks)
```
ditto, in this context it doesn't matter that it's the PodIfaceNameMap

RamLavi commented:
/test pull-kubevirt-unit-test
/test pull-kubevirt-generate
/test pull-kubevirt-build
/test pull-kubevirt-e2e-k8s-1.24-sig-network
/test pull-kubevirt-e2e-kind-1.22-sriov

RamLavi commented:
lint fixed [here](https://github.com/kubevirt/kubevirt/compare/e60ba92d582940f1224316bfe06941952cb493ad..df89582974de973ff61e839ead4102d7ed9f193b)

RamLavi commented:
/test pull-kubevirt-unit-test
/test pull-kubevirt-generate
/test pull-kubevirt-build
/test pull-kubevirt-e2e-k8s-1.24-sig-network
/test pull-kubevirt-e2e-kind-1.22-sriov

RamLavi commented:
@orelmisan fixed the %q [here](https://github.com/kubevirt/kubevirt/compare/df89582974de973ff61e839ead4102d7ed9f193b..d411e3444a8a9f67181ebb3625b1f2d4c1dca393)

orelmisan commented on pkg/network/sriov/sriov.go:
```diff
@@ -19,4 +19,77 @@
 const AliasPrefix = "sriov-"
+
+func CreateNetworkPCIAnnotationValue(networks []virtv1.Network, interfaces []virtv1.Interface, networkStatusAnnotationValue string) string {
+	networkPciMap, err := mapNetworkNameToPCIAddress(networks, interfaces, networkStatusAnnotationValue)
```
nit: I suggest to keep the naming consistent:
`networkPciMap` -> `networkPCIMap`
Or
`mapNetworkNameToPCIAddress` -> `mapNetworkNameToPciAddress`

RamLavi replied:
used networkPCIMap

orelmisan commented on pkg/network/sriov/sriov_test.go:
```diff
@@ -0,0 +1,313 @@
+)
+
+var _ = Describe("SRIOV", func() {
+	nonValidNetworkStatusAnnotationValue := ``
```
nit: any specific reason to use back-ticks here?
Please consider using:
```go
var nonValidNetworkStatusAnnotationValue string
```

Which makes your intention to use an empty string explicit.

RamLavi replied:
back-ticks are used in these strings because they already have "" in the string body. This variable is the exception since it's empty but I think it can stay like that for conformity...

EdDev replied:
It is a nit, but it is indeed odd and the reasoning to have it with "`" is not that strong.
I would even argue that it should be a const.

RamLavi replied:
I re-reviewed it and decided to use the explicit `` instead in the function , and doing that I found a typo [here](https://github.com/kubevirt/kubevirt/pull/8226/files#diff-ef56a1a0d52cf224e4e823c30970416c4ad2bd86b85797f5debc316033b3b669R192) so thanks!

orelmisan commented on pkg/network/sriov/sriov_test.go:
```diff
@@ -0,0 +1,313 @@
+
+var _ = Describe("SRIOV", func() {
+	nonValidNetworkStatusAnnotationValue := ``
+	networkStatusWithOneNetwork := `
```
nit: I suggest naming this variable consistently with the other two: `networkStatusWithOneSRIOVNetwork`

RamLavi replied:
Done

EdDev commented on pkg/virt-launcher/virtwrap/device/hostdevice/sriov/pcipool-network-status_test.go:
```diff
@@ -0,0 +1,89 @@
+			for requestedNetwork, expectedPciAddress := range expectedNetworkToPciMap {
+				By("check pop succeeds to retrieve the PCI Address")
+				address, err := pool.Pop(requestedNetwork)
+				Expect(err).ToNot(HaveOccurred())
```
It is enough to do this:
`Expect(pool.Pop(requestedNetwork)).To(Equal(expectedPciAddress))`

RamLavi replied:
Done

EdDev commented on pkg/virt-launcher/virtwrap/device/hostdevice/sriov/pcipool-network-status_test.go:
```diff
@@ -0,0 +1,89 @@
+				By("check pop fails to retrieve from the network after it was already popped")
+				address, err = pool.Pop(requestedNetwork)
+				Expect(err).To(HaveOccurred())
+				Expect(address).To(Equal(emptyPciAddress))
```
Not needed.

RamLavi replied:
Done

EdDev commented on pkg/virt-launcher/virtwrap/device/hostdevice/sriov/pcipool-network-status_test.go:
```diff
@@ -0,0 +1,89 @@
+			for requestedNetwork, expectedPciAddress := range expectedNetworkToPciMap {
+				address, err := pool.Pop(requestedNetwork)
+				Expect(err).To(HaveOccurred())
+				Expect(address).To(Equal(expectedPciAddress))
```
Not needed

RamLavi replied:
Done

RamLavi commented:
@EdDev fixed [here](https://github.com/kubevirt/kubevirt/compare/d411e3444a8a9f67181ebb3625b1f2d4c1dca393..33e15257749af291ff6acc09979784b329c8967e)

EdDev commented on pkg/virt-controller/watch/vmi.go:
```diff
@@ -345,7 +350,30 @@ func (c *VMIController) execute(key string) error {
+		var err error
+		patchedPod, err = c.clientset.CoreV1().Pods(pod.Namespace).Patch(context.Background(), pod.Name, types.JSONPatchType, patchBytes, v1.PatchOptions{})
+		if err != nil {
+			log.Log.Object(pod).Errorf("failed to sync pod annotations during sync: %s", err)
```
Shouldn't this use `%v`?

RamLavi replied:
fixed

EdDev commented on pkg/virt-controller/watch/vmi.go:
```diff
@@ -638,6 +666,17 @@ func (c *VMIController) updateStatus(vmi *virtv1.VirtualMachineInstance, pod *k8
+func prepareAnnotationsPatchAddOp(key, value string) (string, error) {
+	valueBytes, err := json.Marshal(value)
+	if err != nil {
+		return "", fmt.Errorf("failed to prepare new annotation patchOp for key %s: %q", key, err)
```
I think `%v` needs to be used here (instead of `%q`).

RamLavi replied:
fixed

EdDev commented on pkg/virt-launcher/virtwrap/device/hostdevice/sriov/hostdev.go:
```diff
@@ -34,7 +39,43 @@ import (
+	return CreateHostDevicesFromIfacesAndPool(SRIOVInterfaces, pciAddressPoolWithNetworkStatus)
+}
+
+func newPCIAddressPoolWithNetworkStatusFromFile(path string) (*PCIAddressWithNetworkStatusPool, error) {
```
I think this logic here is not trivial and should have been explained in the function description.
Explicitly, when an error is returned, when no error is returned but also no pool.

A common reader who is not in deep details will have hard time understanding the intention here and the logic of when to fail and when not.

RamLavi replied:
Added function documentation, I agree it's not trivial.

Also temporarily changed it to public to make sure the Len returns 0 in case we want to fallback
```
	It("CHECK CHECK", func() {
		pool, err := sriovhostdev.RAMnewPCIAddressPoolWithNetworkStatusFromFile("/sdf/sdf")

		Expect(err).ToNot(HaveOccurred())
		Expect(pool.Len()).To(BeZero())
	})
```

passed

EdDev commented on pkg/virt-launcher/virtwrap/device/hostdevice/sriov/pcipool_netstatus.go:
```diff
@@ -0,0 +1,66 @@
+	"fmt"
+)
+
+type PCIAddressWithNetworkStatusPool struct {
+	networkPCIMap map[string]string
+}
```
Just an observation: This could have just been `type PCIAddressWithNetworkStatusPool map[string]string`.

RamLavi replied:
and an interesting one at that.

RamLavi commented:
/test pull-kubevirt-unit-test
/test pull-kubevirt-generate
/test pull-kubevirt-build
/test pull-kubevirt-e2e-k8s-1.24-sig-network
/test pull-kubevirt-e2e-kind-1.22-sriov

EdDev reviewed (approved):
Thank you, it looks good!

I had some minor comments inline.

> sriov: Map sriov devices to pci addresses 

I would suggest adding `network, ` in the front for the first commit.

RamLavi commented:
@orelmisan @EdDev fixes [here](https://github.com/kubevirt/kubevirt/compare/fc0fae0b1c3e8a8c9582e8079a895d9cd558d19e..c13661961384078e795be94806e8a27a2f5560fc)

RamLavi commented:
/test pull-kubevirt-code-lint
it's passes locally, prow flake?

EdDev reviewed (approved):
Thank you!

RamLavi commented:
/retest-required

orelmisan commented on pkg/virt-launcher/virtwrap/device/hostdevice/sriov/pcipool_netstatus.go:
```diff
@@ -0,0 +1,66 @@
+func (p *PCIAddressWithNetworkStatusPool) Pop(networkName string) (string, error) {
+	pciAddress, exists := p.networkPCIMap[networkName]
+	if !exists {
+		return "", fmt.Errorf("PCI-Address for SR-IOV network %s not found", networkName)
```
nit:
```suggestion
		return "", fmt.Errorf("PCI-Address for SR-IOV network %q not found", networkName)
```

RamLavi replied:
Done

orelmisan commented on pkg/virt-launcher/virtwrap/device/hostdevice/sriov/pcipool_netstatus.go:
```diff
@@ -0,0 +1,66 @@
+	}
+
+	if pciAddress == "" {
+		return "", fmt.Errorf("failed to allocate SR-IOV PCI address for network %s", networkName)
```
nit:
```suggestion
		return "", fmt.Errorf("failed to allocate SR-IOV PCI address for network %q", networkName)
```

RamLavi replied:
Done

orelmisan commented on pkg/virt-launcher/virtwrap/device/hostdevice/sriov/pcipool_netstatus_test.go:
```diff
@@ -0,0 +1,79 @@
+)
+
+var _ = Describe("SRIOV PCI address pool with network-pci-map", func() {
+	emptyFileBytes := []byte(``)
```
nit:
```suggestion
	var emptyFileBytes []byte
```

RamLavi replied:
Done

orelmisan commented on pkg/virt-launcher/virtwrap/device/hostdevice/sriov/pcipool_netstatus_test.go:
```diff
@@ -0,0 +1,79 @@
+
+var _ = Describe("SRIOV PCI address pool with network-pci-map", func() {
+	emptyFileBytes := []byte(``)
+	emptyNetworkPCIMapBytes := []byte(`{}`)
```
nit:
```suggestion
	emptyNetworkPCIMapBytes := []byte("{}")
```

RamLavi replied:
Done

orelmisan reviewed (commented):
Thank you for the changes @RamLavi.
Just a few nits.
We can change them in a follow-up PR.

AlonaKaplan commented on pkg/virt-controller/watch/vmi.go:
```diff
@@ -1056,6 +1095,18 @@ func (c *VMIController) sync(vmi *virtv1.VirtualMachineInstance, pod *k8sv1.Pod,
 	}
 
 	if !isTempPod(pod) && isPodReady(pod) {
+		if vmispec.SRIOVInterfaceExist(vmi.Spec.Domain.Devices.Interfaces) {
```
Is there a special reason there are no unit tests?

RamLavi replied:
I forgot to add it here, but now I am hitting some issues..

RamLavi replied:
Done

ormergi commented on pkg/virt-controller/watch/vmi.go:
```diff
@@ -1056,6 +1095,18 @@ func (c *VMIController) sync(vmi *virtv1.VirtualMachineInstance, pod *k8sv1.Pod,
+			if err != nil {
+				return &syncErrorImpl{err, FailedPodPatchReason}
+			}
+			*pod = *patchedPod
```
I think that changing it as follows will make the code easier to understand, WDYT?
```
var err error
...
pod, err = c.syncPodAnnotations(pod, newAnnotations)
...
Or
if pod, err := c.syncPodAnnotations(pod, newAnnotations); err != nil {
...
```

RamLavi replied:
I don't think so.
the whole point is for `pod` to adjust to the changes made but NOT change its pointer address. 
patchedPod is a temporary variable that will loose context outside of this scope.

RamLavi replied:
> I think that changing it as follows will make the code easier to understand, WDYT?
> 
> ```
> var err error
> ...
> pod, err = c.syncPodAnnotations(pod, newAnnotations)
> ...
> ```

I think it's better to explicitly say that we are changing the pod to be patched.

RamLavi replied:
> I think that changing it as follows will make the code easier to understand, WDYT?
> 
> ```
> var err error
> ...
> pod, err = c.syncPodAnnotations(pod, newAnnotations)
> ...
> ```

I think this way is more explicit to suggest that the pod object is changed. but I'm kinda on the fence since it's not common practice..

RamLavi commented on pkg/network/sriov/sriov.go:
```diff
@@ -19,4 +19,77 @@
+
+		pciAddress := networkStatusEntry.DeviceInfo.Pci.PciAddress
+		if pciAddress == "" {
+			return nil, fmt.Errorf("failed to associate pci-address to SRIOV interface %q", sriovIface.Name)
```
Done

RamLavi commented on pkg/network/sriov/sriov_suite_test.go:
```diff
@@ -0,0 +1,11 @@
+package sriov_test
```
Added. Thanks!

RamLavi commented on pkg/network/sriov/sriov_test.go:
```diff
@@ -0,0 +1,312 @@
+		Entry("when pod's networkStatus Annotation does not exist",
+			[]virtv1.Network{newMultusNetwork("foo", "default/nad1")},
+			[]virtv1.Interface{newSRIOVInterface("foo")},
+			``,
```
Done

RamLavi commented on pkg/virt-launcher/virtwrap/device/hostdevice/sriov/hostdev.go:
```diff
@@ -34,7 +39,49 @@ import (
 	SRIOVInterfaces := vmispec.FilterSRIOVInterfaces(vmi.Spec.Domain.Devices.Interfaces)
-	return CreateHostDevicesFromIfacesAndPool(SRIOVInterfaces, NewPCIAddressPool(SRIOVInterfaces))
+
+	pciAddressPoolWithNetworkStatus, err := newPCIAddressPoolWithNetworkStatusFromFile(sriov.MountPath + "/" + sriov.VolumePath)
```
Like :)

RamLavi commented:
@orelmisan @AlonaKaplan comments addressed [here](https://github.com/kubevirt/kubevirt/compare/c13661961384078e795be94806e8a27a2f5560fc..e5b70d769f3cac2b069f8789aaa03916b390f09d)

orelmisan reviewed (approved):
Thank you for the changes @RamLavi.

RamLavi commented:
/hold
need to check something in my unit test

AlonaKaplan commented on pkg/virt-launcher/virtwrap/device/hostdevice/sriov/hostdev.go:
```diff
@@ -34,7 +40,49 @@ import (
+// PCI-Address Pool.
+// possible return values are:
+// - file populated - return PCI-Address Pool using the data in file.
+// - file empty post-polling (timeout) - return  err.
```
Can you please explain what is the reason behind the decision to return an error just when the file is empty.

RamLavi replied:
Yes. 
First concern is that we want to keep same behavior like before (fallback to old mapping) in cases such as network-status annotation is not in the pod or the PCI data is not int the annotation (as in old Multus deployed or bug in their annotation). 
Other case when we want to "fallback" is when the volume does not exist (possible bug in downwardAPI) or we error reading it. 
In "fallback" scenario - we will return nil but the outputted `pciAddressPoolWithNetworkStatus.Len()` will be 0, thus causing the fallback to [occur](https://github.com/kubevirt/kubevirt/pull/8226/files/b0f2ba81f832e8ef1de0adf15535495aa093d91d..0b93379dfb2e0ee7f42c41d3fb17159cfed5b3cb#diff-4bf5d16bba47083655ec1792df984d6426e722cc13b1aceb5184218146b56f9cR48).

Second concern is that since downwardAPI sometimes takes time for the data to propagate to the volume inside the pod (sometimes is instant, sometimes 400ms by my measurements) . 
In order to not do a SyncVMI err ping pong until the volume is updated (unnecessarily expensive traffic wise), we introduced a short poll of 1 second and then return error so that the SyncVMI gRPC will re-enqueue (this will also propagate the syncError as an event).

AlonaKaplan replied:
Thanks! I meant adding a comment to the code:)

RamLavi replied:
ah, :) 
Done

AlonaKaplan commented on pkg/virt-launcher/virtwrap/device/hostdevice/sriov/pcipool_netstatus.go:
```diff
@@ -0,0 +1,66 @@
+
+// Pop gets the next PCI address available to a particular SR-IOV network. The
+// function makes sure that the allocated address is not allocated to other networks.
+func (p *PCIAddressWithNetworkStatusPool) Pop(networkName string) (string, error) {
```
I seems the method is used for tests only. Please move it to the tests package.

RamLavi replied:
Pop is the API published [here](https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-launcher/virtwrap/device/hostdevice/hostdev.go#L47) and specifically used [here](https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-launcher/virtwrap/device/hostdevice/hostdev.go#L65), after the devices are created

RamLavi commented:
@AlonaKaplan changes [here](https://github.com/kubevirt/kubevirt/compare/0b93379dfb2e0ee7f42c41d3fb17159cfed5b3cb..d0c4d305ca461f6fa671eb22717cf6237af9d1ff):
- added to the function dodumentation
- enhanced the unit test on the patch commit

AlonaKaplan commented:
Thanks!
/approve

EdDev reviewed (approved):
Thanks!

EdDev commented:
@RamLavi , I think this is also resolving #6351 . Could you ref it as well in the description so it will be added to the issues this PR will close when merged?

RamLavi commented:
/retest-required

RamLavi commented:
/hold cancel

RamLavi commented:
/retest-required

RamLavi commented:
/cherry-pick release-0.53

RamLavi commented:
/cherry-pick release-0.57

RamLavi commented:
/cherry-pick release-0.56
